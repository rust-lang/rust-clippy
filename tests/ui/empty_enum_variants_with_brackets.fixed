//@aux-build:proc_macros.rs
#![deny(clippy::empty_enum_variants_with_brackets)]
#![allow(dead_code)]

extern crate proc_macros;
use proc_macros::{external, with_span};

pub enum PublicTestEnum {
    NonEmptyBraces { x: i32, y: i32 },
    NonEmptyParentheses(i32, i32),
    EmptyBraces, //~ empty_enum_variants_with_brackets
    EmptyParentheses(),
}

enum TestEnum {
    NonEmptyBraces { x: i32, y: i32 },
    NonEmptyParentheses(i32, i32),
    EmptyBraces,     //~ empty_enum_variants_with_brackets
    EmptyParentheses, //~ empty_enum_variants_with_brackets
    AnotherEnum,
}

enum TestEnumWithFeatures {
    NonEmptyBraces {
        #[cfg(feature = "thisisneverenabled")]
        x: i32,
    },
    NonEmptyParentheses(#[cfg(feature = "thisisneverenabled")] i32),
}

external! {
    enum External {
        Foo {},
    }
}

with_span! {
    span
    enum ProcMacro {
        Foo(),
    }
}

macro_rules! m {
    ($($ty:ty),*) => {
        enum Macro {
            Foo($($ty),*),
        }
    }
}

mod issue12551 {
    // Check if the variant is passed as an argument.
    enum EvenOdd {
        Even(),
        Odd(),
        Unknown, //~ empty_enum_variants_with_brackets
    }
    fn even_odd(x: i32) -> EvenOdd {
        (x % 2 == 0).then(EvenOdd::Even).unwrap_or_else(EvenOdd::Odd)
    }

    fn natural_number(x: i32) -> NaturalOrNot {
        (x > 0)
            .then(NaturalOrNot::Natural)
            .unwrap_or_else(NaturalOrNot::NotNatural)
    }
    enum NaturalOrNot {
        Natural(),
        NotNatural(),
        Unknown, //~ empty_enum_variants_with_brackets
    }

    // Check that call sites are fixed.
    enum RedundantParenthesesFunctionCall {
        Parentheses, //~ empty_enum_variants_with_brackets
        NoParentheses,
    }
    fn redundant_parentheses_function_call() {
        let _ = RedundantParenthesesFunctionCall::Parentheses;
        let _ = RedundantParenthesesFunctionCall::NoParentheses;
    }

    fn redundant_parentheses_function_call_2() {
        let _ = RedundantParenthesesFunctionCall2::Parentheses;
        let _ = RedundantParenthesesFunctionCall2::NoParentheses;
    }
    enum RedundantParenthesesFunctionCall2 {
        Parentheses, //~ empty_enum_variants_with_brackets
        NoParentheses,
    }
}

fn main() {}
