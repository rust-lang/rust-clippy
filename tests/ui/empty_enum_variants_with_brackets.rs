#![warn(clippy::empty_enum_variants_with_brackets)]

enum TestEnum {
    // These two enums should not trigger the lint.
    NonEmptyBrace { x: i32, y: i32 },
    NonEmptyParentheses(i32, i32),

    // Enum variant with braces and no data
    EmptyBrace {}, //~ ERROR: enum variant with brackets has empty body

    // Enum variant with parentheses and no data
    EmptyParentheses(), //~ ERROR: enum variant with brackets has empty body
}

enum TestEnumWithFeature {
    // These two enums should not trigger the lint.
    NonEmptyBrace {
        #[cfg(feature = "thisisneverenabled")]
        x: i32,
    },
    NonEmptyParentheses(#[cfg(feature = "thisisneverenabled")] i32),
}

fn main() {}
