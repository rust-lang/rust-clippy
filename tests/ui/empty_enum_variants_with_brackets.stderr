error: non-unit variant contains no fields
  --> tests/ui/empty_enum_variants_with_brackets.rs:11:17
   |
LL |     EmptyBraces {},
   |                 ^^
   |
note: the lint level is defined here
  --> tests/ui/empty_enum_variants_with_brackets.rs:2:9
   |
LL | #![deny(clippy::empty_enum_variants_with_brackets)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: remove the braces

error: non-unit variant contains no fields
  --> tests/ui/empty_enum_variants_with_brackets.rs:18:17
   |
LL |     EmptyBraces {},
   |                 ^^
   |
   = help: remove the braces

error: non-unit variant contains no fields
  --> tests/ui/empty_enum_variants_with_brackets.rs:19:21
   |
LL |     EmptyParentheses(),
   |                     ^^
   |
   = help: remove the parenthesis

error: non-unit variant contains no fields
  --> tests/ui/empty_enum_variants_with_brackets.rs:57:16
   |
LL |         Unknown(),
   |                ^^
   |
   = help: remove the parenthesis

error: non-unit variant contains no fields
  --> tests/ui/empty_enum_variants_with_brackets.rs:71:16
   |
LL |         Unknown(),
   |                ^^
   |
   = help: remove the parenthesis

error: non-unit variant contains no fields
  --> tests/ui/empty_enum_variants_with_brackets.rs:76:20
   |
LL |         Parentheses(),
   |                    ^^
   |
   = help: remove the parenthesis

error: non-unit variant contains no fields
  --> tests/ui/empty_enum_variants_with_brackets.rs:89:20
   |
LL |         Parentheses(),
   |                    ^^
   |
   = help: remove the parenthesis

error: aborting due to 7 previous errors

